Don't transform unpublished entities

cvalarida opened the pull request:
## Description
Like it says on the tin, the idea here is that we won't transform unpublished entities.

## Testing done
Tested locally.

## Screenshots
N/A

Commit:
Don't read unpublished entities

Commit:
transformUnpublished option

Commit:
Use in transformer test script

bkjohnson commented on src/site/stages/build/process-cms-exports/index.js:
```diff
@@ -193,6 +215,7 @@ const entityAssemblerFactory = contentDir => {
         parentFieldName,
         contentDir,
         assembleEntityTree,
+        transformUnpublished,
```
It looks like we're passing this extra boolean argument to `transformEntity` defined in `src/site/stages/build/process-cms-exports/transform.js`. Do you want to add an arg there?

Commit:
Merge branch 'master' into no-unpublished-entities-cv

Commit:
Pass transformUnpublished to the function...

...because that's the smart thing to do...

Commit:
Add transformUnpublished to the JSDoc

Commit:
Fix transformers which were previously using unpublished entities

After not transforming unpublished entities, I ran into some
transformers expecting child entities that weren't guaranteed to be
there.

Commit:
Filter null (unpublished) entities from entity reference arrays

Commit:
Handle circular references in replacePathInData

Commit:
Filter null and "empty" pages

Commit:
Ensure .entity

bkjohnson commented on src/site/stages/build/drupal/assets.js:
```diff
@@ -11,20 +11,28 @@ const PUBLIC_URLS_NO_SCHEME = Object.entries(PUBLIC_URLS).reduce(
+  // which is also in its ancestor tree. When this hapens, this function becomes
+  // infinitely recursive. This check looks for circular references and, when
+  // found, exits early because it's already checked that data.
+  if (ancestors.includes(data)) return data;
```
I'm guessing that this also doesn't have to do with unpublished entities, so can you expand the description a bit? If this would cause infinite recursion, how is it just now being noticed?

cvalarida replied:
Right, good call. As I continued working through the Metalsmith build, getting it to use data from the CMS export, I ran into a `Maximum call stack size exceeded` error on this function because data from the CMS export contained circular references.

This isn't directly related to unpublished entities, but I needed the work from this branch to get this far. I just committed this here instead of to a separate branch. :sweat_smile: Further changes are finding a home in a separate branch. :+1:

bkjohnson reviewed (approved):
It isn't super clear why some of the metalsmith plugins are being changed here, so any more context would be helpful.

cvalarida commented on src/site/stages/build/drupal/metalsmith-drupal.js:
```diff
@@ -44,31 +44,11 @@ const shouldPullDrupal = buildOptions =>
-    nullEntities: 0,
-    emptyEntities: 0,
-  };
+  const pages = contentData.data.nodeQuery.entities.filter(
```
Honestly, I don't know if this has to do with unpublished entities or not. After fixing up the transformers, `yarn build:content --use-cms-exports` made it past that stage and into the rest of `getDrupalContent`, but was running into issues with `addHubIconField` failing because an item in `pages` was null. It's unclear to me why this wasn't failing before, as I'm pretty sure `pages` contained null items previously, but this change ensures that all entries in `pages` are neither null nor 
empty objects `{}`. We were skipping them anyhow, so I figured it was safe to filter them out.
